x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobConfigProvider.java

```diff
@@ -0,0 +1,396 @@
+                } catch (Exception e) {
+                    updatedJobListener.onFailure(e);
+                    return;
+                }
+
+                try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                    XContentBuilder updatedSource = updatedJob.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                    IndexRequest indexRequest = client.prepareIndex(AnomalyDetectorsIndex.configIndexName(),
+                            ElasticsearchMappings.DOC_TYPE, Job.documentId(updatedJob.getId()))
+                            .setSource(updatedSource)
+                            .setVersion(version)
+                            .request();
+
+                    executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest, ActionListener.wrap(
+                            indexResponse -> {
+                                assert indexResponse.getResult() == DocWriteResponse.Result.UPDATED;
+                                updatedJobListener.onResponse(updatedJob);
+                            },
+                            updatedJobListener::onFailure
+                    ));
+
+                } catch (IOException e) {
+                    updatedJobListener.onFailure(
+                            new ElasticsearchParseException("Failed to serialise job with id [" + jobId + "]", e));
+                }
+
+
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                updatedJobListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Expands an expression into the set of matching names. {@code expresssion}
+     * may be a wildcard, a job group, a job ID or a list of those.
+     * If {@code expression} == ALL or * then all job IDs are returned.
+     * Job groups are expanded to all the jobs IDs in that group.
+     *
+     * For example, given a set of names ["foo-1", "foo-2", "bar-1", bar-2"],
+     * expressions resolve follows:
+     * <ul>
+     *     <li>"foo-1" : ["foo-1"]</li>
+     *     <li>"bar-1" : ["bar-1"]</li>
+     *     <li>"foo-1,foo-2" : ["foo-1", "foo-2"]</li>
+     *     <li>"foo-*" : ["foo-1", "foo-2"]</li>
+     *     <li>"*-1" : ["bar-1", "foo-1"]</li>
+     *     <li>"*" : ["bar-1", "bar-2", "foo-1", "foo-2"]</li>
+     *     <li>"_all" : ["bar-1", "bar-2", "foo-1", "foo-2"]</li>
+     * </ul>
+     *
+     * @param expression the expression to resolve
+     * @param allowNoJobs if {@code false}, an error is thrown when no name matches the {@code expression}.
+     *                     This only applies to wild card expressions, if {@code expression} is not a
+     *                     wildcard then setting this true will not suppress the exception
+     * @param listener The expanded job IDs listener
+     * @return the set of matching names
+     */
+    public void expandJobsIds(String expression, boolean allowNoJobs, ActionListener<Set<String>> listener) {
```

Comment:
The behaviour of `allowNoJobs` has changed slightly as it only applies if `expression` is '*' or '_all' whereas before it applied to each token in a comma delineated list `foo-*,bar`.

Given that the option [isn't documented](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-get-job.html) and the new behaviour matches the description in the [API spec](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/src/test/resources/rest-api-spec/api/xpack.ml.get_jobs.json) I can live with the subtle change

Reply:
The intention was that `allow_no_jobs` would work exactly how `allow_no_indices` works for index wildcards - see https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-index.html

In the case of `foo-*,bar`, an exception should always be thrown if `bar` doesn't exist, and should be thrown if `foo-*` doesn't exist if and only if `allow_no_jobs` is `false`.

The intention was that job wildcard expansion would work like index wildcard expansion, and job groups would work like aliases, so that if anybody asked how they work we'd be able to give a simple answer and people who just assumed there would be consistency between different endpoints would be correct.

Reply:
The original behaviour could be replicated, one option is to make a query for each token in `expression` and check the results.

However, maybe this can still be done with a single query. Check the results and throw if a token is not a wildcard and no job with that Id was returned or if the token is a wildcard check for matching job IDs and throw depending on the value of `allow_no_jobs`. How this could work with paging or a scroll search I do not know.

Reply:
> How this could work with paging or a scroll search I do not know

I think we'd have to set up a map of token to boolean before starting, record what we saw during the paging or scroll, then throw errors at the very end.

Reply:
I pushed a change to replicate the old `allow_no_jobs` functionality, it's a little complicated but hopefully the code is self-explanatory (if not it needs a rewrite). This should work with any future scroll search/paging change